### ISSUE TITLE
Performance drops over the course of training...

### ISSUE THREAD

**Issue by alexjc** [KellerJordan/modded-nanogpt#40]
Has anyone else noticed that the average step time increases noticeably over the course of training?  It feels like some kind of resource leakage but I haven't tracked it down yet.  I'm using the latest `pytorch-nightly`.

For testing:
* I've disabled the variable attn_blocksize. (I thought that was the problem initially.)
* I've disabled evaluation and saving models.
* I've tested 2x3090 and 8xA100, both slow down to varying degrees.
* I've tried with and without compilation — both slow down, but when compiled is relatively worse.
* I've tried one or more proc_per_node and it's still there.

I've seen the average step get slower 18% over the course of the whole training. Also, testing different kernels I saw drop in 3% average over 30 steps...

Any ideas?

**Comment by alexjc** [comment]
It seems a lot better with `torch 2.6.0.dev20241130` but still a problem...

What 2.6.0 nightly version is everyone using and did you notice any slowdown over the course of training?

**Comment by KellerJordan** [comment]
this is due to the attention window warmup introduced in the most recent record

**Comment by alexjc** [comment]
I thought so too initially, but I turned that off to try to find the root cause of the problem. As I wrote:
> * I've disabled the variable attn_blocksize.

My current run is fixed at `attn_blocksize=64` and goes from 115019.32 tok/s to 76247.63 tok/s over a few hundred steps even!  I considered it may be heat-related GPU regulation, but that's a bit too extreme.

**Comment by KellerJordan** [comment]
oh. not sure then

**Comment by alexjc** [comment]
I found three things so far:
- Problems with `flex_attention` and the document mask which is much slower. AFAICT, the sparsity is not helping performance at all, and benefits on final score not clear to me.
- A performance cliff on 3090 GPUs when going from `n_embd=512` to `768`. Performance becomes more erratic and degrades over time; maybe torch.compile edge case.
- Variability due to GPUs heating / cooling, because I'm running things for hours on a pair of 3090 GPUs which are more likely to trigger throttling in the hardware than 5-minute runs on H100 GPUs.

Still investigating, especially the first point...

**Comment by alxndrTL** [comment]
Have you tried like one run after another ? And see if at the start of the second run, time per step is the same as the last step of the first run.. If there is run dependence then it is surely hardware-related

**Comment by alexjc** [comment]
I get significant variations when restarting, for example to test new versions and have torch recompile the graph. Sometimes 20% boost sometimes 20% drop on things that should not affect performance.

I agree there is some hardware-level problem, but I also found torch versions to perform differently — it is a nightly release after all...

**Comment by alexjc** [comment]
I installed `torch==2.6.0.dev20241203+cu124` and now the performance is very stable towards the mid- to end of training.  (In the current run it even gets faster.)

> pip install --pre torch==2.6.0.dev20241203+cu124 --index-url https://download.pytorch.org/whl/nightly/cu124 --upgrade

Hardware and other questions of rig setup didn't change.  Note that the torch nightly from 20241205 doesn't work on 3090s.

Going to close this thread now, and stick to this particular package until the official release!
